### PR TITLE
Multi-target .NET 5

### PR DESCRIPTION
**Pull request by JamesNK**
Testing multi-target packages on .NET 5

Soon the client and server will start using .NET 5 specific APIs, e.g. configuring the client to create a connections when there are more than 100 concurrent requests.

### CODE REVIEW

**Comment by JamesNK** [comment]
Interop tests are failing because of https://github.com/grpc/grpc/pull/23523

**Comment by JamesNK** [comment]
Everything is passing apart from interop tests

@jtattermusch When you have time can you help with https://github.com/grpc/grpc/pull/23523. Need to update grpc/grpc so changing the framework doesn't change the path and break interop tests.

File: global.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
   "sdk": {
-    "version": "3.1.300"
+    "version": "5.0.100-rc.1.20367.2"
```
Comment (jtattermusch) [inline global.json:3]: should we explicitly set rollForward ?

File: testassets/InteropTestsClient/InteropTestsClient.csproj [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <TargetFramework>net5.0</TargetFramework>
```
Comment (jtattermusch) [inline testassets/InteropTestsClient/InteropTestsClient.csproj:5]: qq: is switching the  Grpc.Core based interop client to net5.0 required? (I'm assuming this is done for consistency?)

Comment (JamesNK) [reply testassets/InteropTestsClient/InteropTestsClient.csproj:5]: It's just done for consistency.

File: global.json (hunk 1 above)
Comment (JamesNK) [reply global.json:3]: I don't know. I have used that setting.

Is it required? I haven't had trouble with installing newer versions yet.

**Comment by JamesNK** [comment]
@JunTaoLuo This is ready to review

File: examples/Worker/Client/Client.csproj [hunk 3]
```diff
@@ -1,5 +1,9 @@
-﻿<Project Sdk="Microsoft.NET.Sdk.Worker">
-
+﻿<Project Sdk="Microsoft.NET.Sdk.Web">
+  <!--
+    TODO(JamesNK): SDK temporarily changed from Microsoft.NET.Sdk.Worker to Microsoft.NET.Sdk.Web
```
Comment (JunTaoLuo) [inline examples/Worker/Client/Client.csproj:3]: File an issue so we don't forget to do this in the future?

File: test/FunctionalTests/Infrastructure/ExecutionContextLoggingProvider.cs [hunk 4]
```diff
@@ -77,11 +77,15 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
                 // if it is written in the test's execution context.
                 ExecutionContext.Run(_executionContext, s =>
                 {
-                    var timestamp = $"{(DateTimeOffset.UtcNow - _timeStart).TotalSeconds.ToString("N3")}s";
+                    var timestamp = $"{(DateTimeOffset.UtcNow - _timeStart).TotalSeconds:N3}s";
```
Comment (JunTaoLuo) [inline test/FunctionalTests/Infrastructure/ExecutionContextLoggingProvider.cs:80]: Is this a new syntax for specifying the format? I've never seen this before.

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs [hunk 5]
```diff
@@ -387,7 +387,7 @@ public void Initialize(ISystemClock? clock = null)
                 ResponseGrpcEncoding = GrpcProtocolConstants.IdentityGrpcEncoding;
             }
 
-            HttpContext.Response.Headers.Append(GrpcProtocolConstants.MessageEncodingHeader, ResponseGrpcEncoding);
+            HttpContext.Response.Headers[GrpcProtocolConstants.MessageEncodingHeader] = ResponseGrpcEncoding;
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs:390]: This has a functional difference right? One is adding to existing headers, one is overwriting? Is that what we want?

File: src/Grpc.Net.Client/GrpcChannel.cs [hunk 6]
```diff
@@ -16,17 +16,16 @@
...
-using System.Collections.Generic;
-using System.Linq;
-using System.Net.Http;
-using System.Threading;
 using Grpc.Core;
 using Grpc.Net.Client.Internal;
 using Grpc.Net.Compression;
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Logging.Abstractions;
+using System;
```
Comment (JunTaoLuo) [inline src/Grpc.Net.Client/GrpcChannel.cs:24]: Why?

File: src/Grpc.Net.Client/Internal/TelemetryHeaderHandler.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            var currentActivity = Activity.Current;
+            if (currentActivity != null)
+            {
+                InjectHeaders(currentActivity, request);
+            }
+
+            return base.SendAsync(request, cancellationToken);
+        }
+
+        private static void InjectHeaders(Activity currentActivity, HttpRequestMessage request)
```
Comment (JunTaoLuo) [inline src/Grpc.Net.Client/Internal/TelemetryHeaderHandler.cs:55]: This looks right to me, but I'm not too familiar with this so might want someone else to double check.

File: src/Grpc.AspNetCore.Server/Internal/CallHandlers/ServerCallHandlerBase.cs [hunk 8]
```diff
@@ -51,32 +51,12 @@ public Task HandleCallAsync(HttpContext httpContext)
...
-                    return Task.CompletedTask;
-                }
-                else
-                {
-                    GrpcServerLog.UnsupportedRequestContentType(Logger, httpContext.Request.ContentType);
-
-                    GrpcProtocolHelpers.BuildHttpErrorResponse(httpContext.Response, StatusCodes.Status415UnsupportedMediaType, StatusCode.Internal, error);
-                    return Task.CompletedTask;
-                }
+                return ProcessInvalidContentTypeRequest(httpContext, error);
```
Comment (JunTaoLuo) [inline src/Grpc.AspNetCore.Server/Internal/CallHandlers/ServerCallHandlerBase.cs:54]: Why extract this method? Seems unnecessary.

File: test/FunctionalTests/Infrastructure/ExecutionContextLoggingProvider.cs (hunk 4 above)
Comment (JamesNK) [reply test/FunctionalTests/Infrastructure/ExecutionContextLoggingProvider.cs:80]: It's always been possible

File: src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs (hunk 5 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs:390]: Overwriting is correct. It is also slightly faster.

File: src/Grpc.Net.Client/Internal/TelemetryHeaderHandler.cs (hunk 7 above)
Comment (JamesNK) [reply src/Grpc.Net.Client/Internal/TelemetryHeaderHandler.cs:55]: This is a simplified version of the diagnostics handler in .NET Core that only applies the header. Should be fine.

File: src/Grpc.AspNetCore.Server/Internal/CallHandlers/ServerCallHandlerBase.cs (hunk 8 above)
Comment (JamesNK) [reply src/Grpc.AspNetCore.Server/Internal/CallHandlers/ServerCallHandlerBase.cs:54]: Just splitting the existing method up into smaller parts

**Pull request merged**